plugins/glustershd/rest.go

```diff
@@ -187,3 +190,280 @@ func glustershDisableHandler(w http.ResponseWriter, r *http.Request) {
+
+}
+
+func granularHealEnableHandler(w http.ResponseWriter, r *http.Request) {
+	p := mux.Vars(r)
+	volname := p["name"]
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	//validate volume name
+	v, err := volume.GetVolume(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusNotFound, err)
+		return
+	}
+
+	// Store volinfo before changing Granular Heal Entry.
+	tmp := *v
+	oldvolinfo := &tmp
+
+	// validate volume type
+	if !isVolReplicate(v.Type) {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume Type not supported")
+		return
+	}
+
+	if v.State != volume.VolStarted {
+		restutils.SendHTTPError(ctx, w, http.StatusBadRequest, "Volume should be in started state.")
+		return
+	}
+
+	txn := transaction.NewTxn(ctx)
+	defer txn.Cleanup()
+
+	//Lock on Volume Name
+	lock, unlock, err := transaction.CreateLockSteps(volname)
+	if err != nil {
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+
+		return
+	}
+
+	v.GranularHealEntryEnabled = true
+
+	txn.Nodes = v.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc:   "granular-heal.Enable",
+			Nodes:    []uuid.UUID{gdctx.MyUUID},
+			UndoFunc: "granular-heal.Undo",
+		},
+		{
+			DoFunc: "vol-option.UpdateVolinfo",
+			Nodes:  txn.Nodes,
+		},
+		unlock,
+	}
+
+	if err := txn.Ctx.Set("volinfo", v); err != nil {
+		logger.WithError(err).Error("failed to set volinfo in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	if err := txn.Ctx.Set("oldvolinfo", oldvolinfo); err != nil {
+		logger.WithError(err).Error("failed to set oldvolinfo in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	if err := txn.Ctx.Set("volname", volname); err != nil {
+		logger.WithError(err).Error("failed to set volname in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	var options []string
+	glusterdSockpath := path.Join(config.GetString("rundir"), "glusterd2.socket")
+	options = append(options, "granular-entry-heal-op", "glusterd-sock", glusterdSockpath)
+	if err := txn.Ctx.Set("option", options); err != nil {
+		logger.WithError(err).Error("failed to set volname in transaction context")
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	err = txn.Do()
+	if err != nil {
+		logger.WithFields(log.Fields{
+			"error":   err.Error(),
+			"volname": volname,
+		}).Error("failed to enable Granular Entry Heal Option")
+
+		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		return
+	}
+
+	restutils.SendHTTPResponse(ctx, w, http.StatusOK, nil)
+}
+
+func granularHealDisableHandler(w http.ResponseWriter, r *http.Request) {
+	p := mux.Vars(r)
+	volname := p["name"]
+
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+
+	//validate volume name
```

Comment:
Same comment here - do most of these validations under lock.

Reply:
Leaving this out for now as discussed.

Reply:
No, not this one. I was referring to removing contents of SocketFile().

Reply:
But we discussed that for now I should leave the validations part as it is and not include them in a step function.

Reply:
@vpandey-RH It can be moved. There's no technical reasons or dependency to not to move. The only problem with moving is that the stepfunction cannot specify what HTTP status to return to client. I am okay with that.